crates/matrix-sdk-base/src/store/memory_store.rs

```diff
@@ -700,17 +702,59 @@ impl StateStore for MemoryStore {
         Ok(self.custom.write().unwrap().remove(key))
     }
 
-    // The in-memory store doesn't cache media
-    async fn add_media_content(&self, _request: &MediaRequest, _data: Vec<u8>) -> Result<()> {
+    async fn add_media_content(&self, request: &MediaRequest, data: Vec<u8>) -> Result<()> {
+        // Avoid duplication. Let's try to remove it first.
+        self.remove_media_content(request).await?;
+        // Now, let's add it.
+        self.media.write().unwrap().push((request.uri().to_owned(), request.unique_key(), data));
+
         Ok(())
     }
-    async fn get_media_content(&self, _request: &MediaRequest) -> Result<Option<Vec<u8>>> {
+
+    async fn get_media_content(&self, request: &MediaRequest) -> Result<Option<Vec<u8>>> {
+        let media = self.media.read().unwrap();
+        let expected_key = request.unique_key();
+
+        for (_media_uri, media_key, media_content) in media.iter() {
+            if media_key == &expected_key {
+                return Ok(Some(media_content.to_owned()));
+            }
+        }
+
         Ok(None)
     }
-    async fn remove_media_content(&self, _request: &MediaRequest) -> Result<()> {
+
+    async fn remove_media_content(&self, request: &MediaRequest) -> Result<()> {
+        let mut media = self.media.write().unwrap();
+        let expected_key = request.unique_key();
+        let Some(index) = media
+            .iter()
+            .position(|(_media_uri, media_key, _media_content)| media_key == &expected_key)
+        else {
+            return Ok(());
+        };
+
+        media.remove(index);
+
         Ok(())
     }
-    async fn remove_media_content_for_uri(&self, _uri: &MxcUri) -> Result<()> {
+
+    async fn remove_media_content_for_uri(&self, uri: &MxcUri) -> Result<()> {
+        let mut media = self.media.write().unwrap();
+        let expected_key = uri.to_owned();
+        let positions = media
+            .iter()
+            .enumerate()
+            .filter_map(|(position, (media_uri, _media_key, _media_content))| {
+                (media_uri == &expected_key).then_some(position)
+            })
+            .collect::<Vec<_>>();
```

Comment:
Do we need to collect() here, since we're using the vec consumed into an iterator immediately thereafter?